File: src/test/java/org/algohub/engine/judge/JavaEngineTest.java
package org.algohub.engine.judge;

import com.google.common.base.Charsets;
import com.google.common.io.Resources;
import junitparams.JUnitParamsRunner;
import junitparams.Parameters;
import org.algohub.engine.ObjectMapperInstance;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.fail;

@RunWith(JUnitParamsRunner.class)
public class JavaEngineTest {

    private static void judgeSolution(final String problemId, final String solutionId, StatusCode statusCode) {
        try {
            final String questionStr =
                    Resources.toString(Resources.getResource(problemId + ".json"), Charsets.UTF_8);
            final Problem problem =
                    ObjectMapperInstance.INSTANCE.getINSTANCE().readValue(questionStr, Problem.class);
            final String sourceCode =
                    Resources.toString(Resources.getResource(solutionId + ".java"), Charsets.UTF_8);

            final JudgeResult result = JudgeEngine.INSTANCE.judge(problem, sourceCode);

            assertThat(result.getStatusCode()).isEqualTo(statusCode.toString());
        } catch (IOException e) {
            fail(e.getMessage());
        }
    }

    @Test
    @Parameters({
            "2-sum, TwoSum",
            "fib, FibFast",
            "stoi, MyStoi",
            "word-ladder, WordLadder"
    })
    public void acceptsCorrectSolution(String problemId, String solutionId) throws Exception {
        judgeSolution(problemId, solutionId, StatusCode.ACCEPTED);
    }

    @Test
    public void failsWithCompilationErrorWhenSourceCodeDoesNotCompile() throws Exception {
        judgeSolution("fib", "FibBroken", StatusCode.COMPILE_ERROR);
    }

    @Test
    public void returnsWrongAnswerForUncorrectSolution() throws Exception {
        judgeSolution("fib", "FibWrongAnswer", StatusCode.WRONG_ANSWER);
    }
}

File: src/test/kotlin/org/algohub/engine/judge/JavaEngineTest.kt
package org.algohub.engine.judge

import com.google.common.base.Charsets
import com.google.common.io.Resources
import junitparams.JUnitParamsRunner
import junitparams.Parameters
import org.algohub.engine.ObjectMapperInstance
import org.junit.Test
import org.junit.runner.RunWith

import java.io.IOException

import org.assertj.core.api.Assertions.assertThat
import org.junit.Assert.fail

@RunWith(JUnitParamsRunner::class)
class JavaEngineTest {

    private fun judgeSolution(problemId: String, solutionId: String, statusCode: StatusCode) {
        try {
            val questionStr = Resources.toString(Resources.getResource("$problemId.json"), Charsets.UTF_8)
            val problem = ObjectMapperInstance.INSTANCE.readValue(questionStr, Problem::class.java)
            val sourceCode = Resources.toString(Resources.getResource("$solutionId.java"), Charsets.UTF_8)

            val result = JudgeEngine.judge(problem, sourceCode)

            assertThat(result.statusCode).isEqualTo(statusCode.toString())
        } catch (e: IOException) {
            fail(e.message)
        }

    }

    @Test
    @Parameters("2-sum, TwoSum", "fib, FibFast", "stoi, MyStoi", "word-ladder, WordLadder")
    @Throws(Exception::class)
    fun acceptsCorrectSolution(problemId: String, solutionId: String) {
        judgeSolution(problemId, solutionId, StatusCode.ACCEPTED)
    }

    @Test
    @Throws(Exception::class)
    fun failsWithCompilationErrorWhenSourceCodeDoesNotCompile() {
        judgeSolution("fib", "FibBroken", StatusCode.COMPILE_ERROR)
    }

    @Test
    @Throws(Exception::class)
    fun returnsWrongAnswerForIncorrectSolution() {
        judgeSolution("fib", "FibWrongAnswer", StatusCode.WRONG_ANSWER)
    }
}
